### PR TITLE
[Feat] Support update weights without restart server

### PR DESCRIPTION
**Pull request by shanyu-sys**
## Motivation
Support updating the model weights without relaunching the server.
#634 

## Modification
* Support post a "/update_weights" request in `server.py`
* Add `update_weights` methods in `tokenizer_manager`, `tp_worker` and `model_runner`
* Add unit test in `test_update_weights.py`

## Checklist

- [x] Before submitting a PR for review, make sure it has passed verification in your local development environment **at least**.
- [x] Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
- [x] Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
- [ ] Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

File: test/srt/test_update_weights.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        new_model_path = "meta-llama/Meta-Llama-3.1-8B"
+        self.run_update_weights(new_model_path)
+
+        updated_model_path = self.get_model_info()
+        print(f"updated_model_path: {updated_model_path}")
+        assert updated_model_path == new_model_path
+        assert updated_model_path != origin_model_path
+
+        updated_response = self.run_decode()
+        assert origin_response[:32] != updated_response[:32]
```
Comment (Ying1123) [inline test/srt/test_update_weights.py:78]: To test if the model can be updated correctly. Could you add one more step of `self.run_update_weights(origin_model_path)`, and later `assert origin_response[:32] == updated_response[:32]`?

**Review by Ying1123 — APPROVED** [review]

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -204,6 +207,91 @@ def load_model(self):
...
+                model = model_load_weights(self.model, iter)
+            except Exception as e:
+                logger.error(
+                    f"Failed to update weights: {e}. \n Rolling back to original weights"
+                )
+                del iter
+                gc.collect()
+                iter = get_weight_iter(self.vllm_model_config)
+                self.model = model_load_weights(self.model, iter)
+                return False, "Failed to update model"
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/model_runner.py]: add `{e}` to the return message

File: python/sglang/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -204,6 +207,91 @@ def load_model(self):
...
+                    with device_loading_context(module, target_device):
+                        quant_method.process_weights_after_loading(module)
+            return model
+
+        with set_default_torch_dtype(vllm_model_config.dtype):
+            try:
+                iter = get_weight_iter(vllm_model_config)
+            except Exception as e:
+                logger.error(f"Failed to get weights iterator: {e}")
+                return False, "Failed to update model"
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/model_runner.py]: add `{e}` to the return message

File: python/sglang/srt/managers/tp_worker.py [hunk 4]
```diff
@@ -798,6 +803,14 @@ def abort_request(self, recv_req):
                     req.finished_reason = FINISH_ABORT()
                     break
 
+    def update_weights(self, recv_req):
+        success, message = self.model_runner.update_weights(
+            recv_req.model_path, recv_req.load_format
+        )
+        if success:
+            self.flush_cache()
```
Comment (merrymercy) [inline python/sglang/srt/managers/tp_worker.py]: currently, `flush_cache` only prints a warning if there are some errors.
1. add a return value to `flush_cache` to indicate whether it is successful
2. assert the return value here

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/model_executor/model_runner.py (hunk 3 above)
Comment (shanyu-sys) [reply python/sglang/srt/model_executor/model_runner.py]: Done

File: python/sglang/srt/managers/tp_worker.py (hunk 4 above)
Comment (shanyu-sys) [reply python/sglang/srt/managers/tp_worker.py]: Done

File: test/srt/test_update_weights.py (hunk 1 above)
Comment (shanyu-sys) [reply test/srt/test_update_weights.py:78]: Done

File: python/sglang/srt/model_executor/model_runner.py (hunk 2 above)
Comment (shanyu-sys) [reply python/sglang/srt/model_executor/model_runner.py]: Done

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 5]
```diff
@@ -146,6 +152,9 @@ async def generate_request(
         if self.to_create_loop:
             self.create_handle_loop()
 
+        while self.model_update_lock.locked():
+            await asyncio.sleep(0.1)
```
Comment (zhyncs) [inline python/sglang/srt/managers/tokenizer_manager.py]:
```suggestion
            await asyncio.sleep(0)
```
Sleep 0.1 doesn't have any actual meaning. Essentially, it makes the current coroutine be moved to the end of the event loop, waiting for the next round of the loop to continue execution.

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 6]
```diff
@@ -513,6 +522,30 @@ def flush_cache(self):
...
+
+        # default the load format to the server_args
+        if obj.load_format is None:
+            obj.load_format = self.server_args.load_format
+
+        if not self.model_update_lock.locked():
+            async with self.model_update_lock:
+                # wait for the previous generation requests to finish
+                while len(self.rid_to_state) > 0:
+                    await asyncio.sleep(0.1)
```
Comment (zhyncs) [inline python/sglang/srt/managers/tokenizer_manager.py]: same as mentioned above

File: python/sglang/srt/model_executor/model_runner.py [hunk 7]
```diff
@@ -206,6 +209,91 @@ def load_model(self):
...
+
+        self.model = model
+        self.server_args.model_path = model_path
+        self.server_args.load_format = load_format
+        self.vllm_model_config = vllm_model_config
+        self.load_config = load_config
+        self.model_config.path = model_path
+
+        logger.info(f"[gpu={self.gpu_id}] Update weights end.")
+        return True, "Updating model weights succeeded"
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/model_runner.py]:
```suggestion
        return True, "Succeeded to update model weights"
```

File: python/sglang/srt/managers/tokenizer_manager.py (hunk 5 above)
Comment (shanyu-sys) [reply python/sglang/srt/managers/tokenizer_manager.py]: That makes sense. Thanks!

**Pull request merged**